buildengine/build_go_test.go

```diff
@@ -245,3 +245,74 @@ func TestGoModVersion(t *testing.T) {
 	}
 	testBuild(t, bctx, fmt.Sprintf("go version %q is not recent enough for this module, needs minimum version \"9000.1.1\"", runtime.Version()[2:]), []assertion{})
 }
+
+func TestGeneratedTypeRegistry(t *testing.T) {
+	sch := &schema.Schema{
+		Modules: []*schema.Module{
+			{Name: "another", Decls: []schema.Decl{
+				&schema.Enum{
+					Name:   "TypeEnum",
+					Export: true,
+					Variants: []*schema.EnumVariant{
+						{Name: "A", Value: &schema.TypeValue{Value: &schema.Int{}}},
+						{Name: "B", Value: &schema.TypeValue{Value: &schema.String{}}},
+					},
+				},
+			}},
+		},
+	}
+	expected := `// Code generated by FTL. DO NOT EDIT.
+package main
+
+import (
+	"context"
+	"reflect"
+
+	"github.com/TBD54566975/ftl/common/plugin"
+	"github.com/TBD54566975/ftl/go-runtime/server"
+	"github.com/TBD54566975/ftl/backend/protos/xyz/block/ftl/v1/ftlv1connect"
+	"github.com/TBD54566975/ftl/go-runtime/ftl"
+
+	"ftl/other"
+	"ftl/another"
+)
+
+func main() {
+	verbConstructor := server.NewUserVerbServer("other",
+		server.HandleCall(other.Echo),
+	)
+	ctx := context.Background()
+	typeRegistry := ftl.NewTypeRegistry()
+	typeRegistry.RegisterSumType(reflect.TypeFor[other.D](), map[string]reflect.Type{
+		"Bool": reflect.TypeFor[other.Bool](),
+		"Bytes": reflect.TypeFor[other.Bytes](),
+		"Float": reflect.TypeFor[other.Float](),
+		"Int": reflect.TypeFor[other.Int](),
```

Comment:
<s>I don't think these should be imported from `other` should they? The builtins should be coming from `ftl/builtin`.</s>

Oh haha! I didn't realise they were part of a sum type!

Reply:
Wait...that shouldn't be possible - the builtin types should be reserved words for top-level decls. Have a look at the checks in `Config` et al

Reply:
do we need to add validation for this?

in the schema, "Int", etc. will be the enum variant name. we're not exporting a data decl with this name, if that makes a difference

Reply:
Ah I see, they're only also exported as top-level decls if they're data structs. Makes sense, disregard!